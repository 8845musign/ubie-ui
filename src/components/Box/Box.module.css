.box {
  box-sizing: border-box;
  display: block;
  padding: var(--padding-top) var(--padding-right) var(--padding-bottom) var(--padding-left);
  margin: var(--margin-top) var(--margin-right) var(--margin-bottom) var(--margin-left);
  border-radius: var(--radius);
  font-size: var(--text-size);
  color: var(--text-color);
  line-height: var(--text-leading);
  font-weight: var(--text-bold);
}

.box.backgroundColorPrimary {
  background-color: var(--color-background-primary);
}

.box.backgroundColorPrimaryDarken {
  background-color: var(--color-background-primary-darken);
}

.box.backgroundColorAccent {
  background-color: var(--color-background-accent);
}

.box.backgroundColorAccentDarken {
  background-color: var(--color-background-accent-darken);
}

.box.backgroundColorAlert {
  background-color: var(--color-background-alert);
}

.box.backgroundColorGray {
  background-color: var(--color-background-gray);
}

.box.backgroundColorWhite {
  background-color: var(--color-background-white);
}

.box.borderGray {
  border: 1px solid var(--color-border);
}

.box.borderGrayThick {
  border: 2px solid var(--color-border);
}

.box.borderPrimary {
  border: 1px solid var(--color-primary);
}

.box.borderPrimaryThick {
  border: 2px solid var(--color-primary);
}

.box.widthFull {
  width: 100%;
}

.box.textBold {
  font-weight: bold;
}

.box.textNormal {
  font-weight: normal;
}

.box.textCenter {
  text-align: center;
}

.box.textLeft {
  text-align: left;
}

.box.textRight {
  text-align: right;
}
