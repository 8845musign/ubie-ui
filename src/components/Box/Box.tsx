'use client';

import clsx from 'clsx';
import styles from './Box.module.css';
import {} from '../../types/style';
import { paddingVariables, marginVariables, radiusVariables } from '../../utils/style';
import { HTMLTagname } from '../../utils/types';
import type { CustomDataAttributeProps } from '../../types/attributes';
import type { PaddingProps, MarginProps, RadiusProp, BackgroundColor } from '../../types/style';
import type { FC, PropsWithChildren } from 'react';

type Props = {
  /**
   * レンダリングされるHTML要素
   * @default div
   */
  as?: HTMLTagname;
  /**
   * 背景色
   */
  backgroundColor?: BackgroundColor;
  /**
   * ボーダーの種類
   */
  border?: 'gray' | 'grayThick' | 'primary' | 'primaryThick';
  /**
   * 幅を指定。他のスタイルの影響を受け、幅が100%とならない場合にのみ使用
   */
  width?: 'full';
} & PaddingProps &
  MarginProps &
  RadiusProp &
  CustomDataAttributeProps;

const capitalize = (str: string) => str.charAt(0).toUpperCase() + str.slice(1);

export const Box: FC<PropsWithChildren<Props>> = ({
  as: BoxCopmonent = 'div',
  children,
  pt,
  pr,
  pb,
  pl,
  mt,
  mr,
  mb,
  ml,
  radius,
  backgroundColor,
  border,
  width,
  ...props
}) => {
  return (
    <BoxCopmonent
      className={clsx([
        styles.box,
        backgroundColor && styles[`backgroundColor${capitalize(backgroundColor)}`],
        border && styles[`border${capitalize(border)}`],
        width && styles.widthFull,
      ])}
      style={{
        ...paddingVariables({
          pt,
          pr,
          pb,
          pl,
        }),
        ...marginVariables({
          mt,
          mr,
          mb,
          ml,
        }),
        ...radiusVariables(radius),
      }}
      {...props}
    >
      {children}
    </BoxCopmonent>
  );
};
