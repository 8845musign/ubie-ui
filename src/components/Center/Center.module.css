.center {
  max-width: var(--max-width);
  margin-inline: auto;
  padding: var(--padding-top) var(--padding-right) var(--padding-bottom) var(--padding-left);
}

.center.textCenter {
  text-align: center;
}

.center.childrenCenter {
  display: flex;
  flex-direction: column;
  align-items: center;
}
