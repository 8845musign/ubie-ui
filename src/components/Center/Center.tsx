'use client';

import clsx from 'clsx';
import { isValidElement, cloneElement } from 'react';
import styles from './Center.module.css';
import { createSpacingVariableFromKey, paddingVariables } from '../../utils/style';
import { HTMLTagname } from '../../utils/types';
import { Box } from '../Box/Box';
import type { MarginYProps, PaddingProps } from '../../types/style';
import type { FC, PropsWithChildren, CSSProperties, ComponentType, ReactElement, ReactNode } from 'react';

type Props = {
  /**
   * レンダリングされるHTML要素
   * @default div
   */
  as?: HTMLTagname | ReactElement<ComponentType<typeof Box>>;
  /**
   * 内包するテキストを中央に配置。設定は継承され、子孫要素にも影響します
   * @default false
   */
  textCenter?: boolean;
  /**
   * 子要素を中央に配置。孫要素には影響しません
   * @default false
   */
  childrenCenter?: boolean;
  /**
   * 横幅の最大値
   * @default none
   */
  maxWidth?: string;
  /**
   * HTMLのID属性の値
   */
  id?: string;
} & MarginYProps &
  PaddingProps;

export const Center: FC<PropsWithChildren<Props>> = ({
  as: CenterCopmonent = 'div',
  children,
  pt,
  pr,
  pb,
  pl,
  mt,
  mb,
  textCenter,
  childrenCenter,
  id,
  maxWidth = 'none',
}) => {
  /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
  const createElement = (props: any, children: ReactNode) => {
    if (isValidElement(CenterCopmonent)) {
      return (
          <div {...props}>
            {cloneElement(CenterCopmonent, CenterCopmonent.props, children)}
          </div>
      )
    } else {
      return <CenterCopmonent {...props}>{children}</CenterCopmonent>;
    }
};

  return createElement(
    {
      id,
      className: clsx(styles.center, textCenter && styles.textCenter, childrenCenter && styles.childrenCenter),
      style: {
        '--max-width': maxWidth,
        ...paddingVariables({
          pt,
          pr,
          pb,
          pl,
        }),
        ...{
          '--margin-top': mt ? createSpacingVariableFromKey(mt) : 0,
          '--margin-bottom': mb ? createSpacingVariableFromKey(mb) : 0,
        },
      } as CSSProperties,
    },
    children,
  );
};
