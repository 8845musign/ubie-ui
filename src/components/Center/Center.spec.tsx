import { render, screen } from '@testing-library/react';
import { createRef } from 'react';
import { Center } from './Center';
import { Box } from '../Box/Box';

describe('<Center>', () => {
  it('has a vertical margins of through my prop', () => {
    render(
      <Center my="xxs" data-testid="box">
        Test
      </Center>,
    );
    const div = screen.getByTestId('box');

    expect(div).toHaveStyle('--margin-top: var(--size-spacing-xxs)');
    expect(div).toHaveStyle('--margin-bottom: var(--size-spacing-xxs)');
  });

  it('has all margins through individual margin props', () => {
    render(
      <Center mt="xs" mb="sm" data-testid="box">
        Test
      </Center>,
    );
    const div = screen.getByTestId('box');

    expect(div).toHaveStyle('--margin-top: var(--size-spacing-xs)');
    expect(div).toHaveStyle('--margin-bottom: var(--size-spacing-sm)');
  });

  it('gives priority to individual margin props', () => {
    render(
      <Center my="xxs" mt="xs" mb="sm" data-testid="box">
        Test
      </Center>,
    );
    const div = screen.getByTestId('box');

    expect(div).toHaveStyle('--margin-top: var(--size-spacing-xs)');
    expect(div).toHaveStyle('--margin-bottom: var(--size-spacing-sm)');
  });

  it('has all the paddings through p prop', () => {
    render(
      <Center p="xxs" data-testid="box">
        Test
      </Center>,
    );
    const div = screen.getByTestId('box');

    expect(div).toHaveStyle('--padding-top: var(--size-spacing-xxs)');
    expect(div).toHaveStyle('--padding-right: var(--size-spacing-xxs)');
    expect(div).toHaveStyle('--padding-bottom: var(--size-spacing-xxs)');
    expect(div).toHaveStyle('--padding-left: var(--size-spacing-xxs)');
  });

  it('has a horizontal paddings through px prop', () => {
    render(
      <Center px="xxs" data-testid="box">
        Test
      </Center>,
    );
    const div = screen.getByTestId('box');

    expect(div).toHaveStyle('--padding-top: 0');
    expect(div).toHaveStyle('--padding-right: var(--size-spacing-xxs)');
    expect(div).toHaveStyle('--padding-bottom: 0');
    expect(div).toHaveStyle('--padding-left: var(--size-spacing-xxs)');
  });

  it('has a vertical paddings of through px prop', () => {
    render(
      <Center py="xxs" data-testid="box">
        Test
      </Center>,
    );
    const div = screen.getByTestId('box');

    expect(div).toHaveStyle('--padding-top: var(--size-spacing-xxs)');
    expect(div).toHaveStyle('--padding-right: 0');
    expect(div).toHaveStyle('--padding-bottom: var(--size-spacing-xxs)');
    expect(div).toHaveStyle('--padding-left: 0');
  });

  it('has all paddings through individual padding props', () => {
    render(
      <Center pt="xxs" pr="xs" pb="sm" pl="md" data-testid="box">
        Test
      </Center>,
    );
    const div = screen.getByTestId('box');

    expect(div).toHaveStyle('--padding-top: var(--size-spacing-xxs)');
    expect(div).toHaveStyle('--padding-right: var(--size-spacing-xs)');
    expect(div).toHaveStyle('--padding-bottom: var(--size-spacing-sm)');
    expect(div).toHaveStyle('--padding-left: var(--size-spacing-md)');
  });

  it('gives priority to individual padding props', () => {
    render(
      <Center p="xxs" px="xs" py="xs" pt="sm" pr="md" pb="lg" pl="xl" data-testid="box">
        Test
      </Center>,
    );
    const div = screen.getByTestId('box');

    expect(div).toHaveStyle('--padding-top: var(--size-spacing-sm)');
    expect(div).toHaveStyle('--padding-right: var(--size-spacing-md)');
    expect(div).toHaveStyle('--padding-bottom: var(--size-spacing-lg)');
    expect(div).toHaveStyle('--padding-left: var(--size-spacing-xl)');
  });

  it('receives width', () => {
    render(
      <Center width="100px" data-testid="flex-item">
        Test
      </Center>,
    );
    const div = screen.getByTestId('flex-item');

    expect(div).toHaveStyle('--width: 100px');
  });

  it('receives max-width', () => {
    render(
      <Center maxWidth="100px" data-testid="flex-item">
        Test
      </Center>,
    );
    const div = screen.getByTestId('flex-item');

    expect(div).toHaveStyle('--max-width: 100px');
  });

  it('receives min-width', () => {
    render(
      <Center minWidth="100px" data-testid="flex-item">
        Test
      </Center>,
    );
    const div = screen.getByTestId('flex-item');

    expect(div).toHaveStyle('--min-width: 100px');
  });

  test('receives data attribute', async () => {
    render(<Center data-testid="center">test</Center>);

    const center = await screen.findByTestId('center');

    expect(center).toBeInTheDocument();
  });

  test('receives ref', async () => {
    const ref = createRef<HTMLDivElement>();

    render(
      <Center ref={ref} data-testid="center">
        test
      </Center>,
    );

    expect(ref.current).not.toBeNull();
    expect(ref.current?.tagName).toBe('DIV');
  });

  test('receives ref', async () => {
    const ref = createRef<HTMLDivElement>();

    render(
      <Center ref={ref} data-testid="center">
        test
      </Center>,
    );

    expect(ref.current).not.toBeNull();
    expect(ref.current?.tagName).toBe('DIV');
  });

  test('changes element to be rendered by as prop', async () => {
    const ref = createRef<HTMLDivElement>();

    render(
      <Center ref={ref} data-testid="center" as="span">
        test
      </Center>,
    );

    expect(ref.current).not.toBeNull();
    expect(ref.current?.tagName).toBe('SPAN');
  });

  test('changes element to be rendered by as prop2', async () => {
    const ref = createRef<HTMLFieldSetElement>();

    render(
      <Center as="fieldset" ref={ref} form="some-form">
        test
      </Center>,
    );

    expect(ref.current).not.toBeNull();
    expect(ref.current?.tagName).toBe('FIELDSET');
    expect(ref.current?.getAttribute('form')).toBe('some-form');
  });

  describe('Snapshot Test', () => {
    it('changes rendered elements', () => {
      render(
        <Center as="section" p="md" data-testid="center">
          Test
        </Center>,
      );
      const center = screen.getByTestId('center');
      expect(center).toMatchSnapshot();
    });

    it('changes the rendered component', () => {
      render(
        <Center p="md" as={<Box backgroundColor="gray" />} data-testid="center">
          Test
        </Center>,
      );
      const center = screen.getByTestId('center');
      expect(center).toMatchSnapshot();
    });
  });
});
