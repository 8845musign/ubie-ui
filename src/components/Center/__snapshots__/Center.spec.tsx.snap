// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Center> > Snapshot Test > changes rendered elements 1`] = `
<label
  class="_center_4c866a"
  data-testid="center"
  for="some-id"
  style="--max-width: none; --padding-top: var(--size-spacing-md); --padding-right: var(--size-spacing-md); --padding-bottom: var(--size-spacing-md); --padding-left: var(--size-spacing-md); --margin-top: 0; --margin-right: 0; --margin-bottom: 0; --margin-left: 0; --width: auto; --min-width: auto;"
>
  Test
</label>
`;

exports[`<Center> > Snapshot Test > changes the rendered component 1`] = `
<div
  class="_center_4c866a"
  data-testid="center"
  style="--max-width: none; --padding-top: var(--size-spacing-md); --padding-right: var(--size-spacing-md); --padding-bottom: var(--size-spacing-md); --padding-left: var(--size-spacing-md); --margin-top: 0; --margin-right: 0; --margin-bottom: 0; --margin-left: 0; --width: auto; --min-width: auto;"
>
  <div
    class="_box_77a981 _backgroundColorGray_77a981"
    style="--padding-top: 0; --padding-right: 0; --padding-bottom: 0; --padding-left: 0; --margin-top: 0; --margin-right: 0; --margin-bottom: 0; --margin-left: 0; --radius: 0; --text-color: inherit; --width: auto; --min-width: auto; --max-width: none;"
  >
    Test
  </div>
</div>
`;
